stdlib/mutex.ml

```diff
@@ -18,3 +18,14 @@ external create: unit -> t = "caml_ml_mutex_new"
 external lock: t -> unit = "caml_ml_mutex_lock"
 external try_lock: t -> bool = "caml_ml_mutex_try_lock"
 external unlock: t -> unit = "caml_ml_mutex_unlock"
+
+(* cannot inline, otherwise flambda might move code around. *)
+let[@inline never] with_locked m f =
+  lock m;
+  match f() with
+  | x ->
+    unlock m; x
+  | exception e ->
+    (* NOTE: [unlock] does not poll for asynchronous exceptions *)
+    unlock m;
+    raise e
```

Comment:
I think that you should use `reraise` instead of `raise`, and that this should guarantee that backtraces are preserved as you originally intended. (It might be that `raise` in fact suffices to get the right backtrace, but using `reraise` explicitly is more robust.) Could you do the change, which might require re-exporting the `%reraise` external locally, and test on your end that backtraces are indeed preserved as expected?

Reply:
I haven't seen `%reraise` bound anywhere in the stdlib, where is it declared and how is it used?

`rg '%reraise` only gives me occurrences in the compiler itself and in the testsuite.

Reply:
We added `reraise` for situations where power users want precise control over their backtraces, before `raise_with_backtrace` was introduced. It is not bound in the standard library because we don't have a clear model of how to expose/specify backtrace stuff right now, so I don't know the "right" way to do it. I guess we could have defined `Printexc.reraise`, but didn't want to make this module even more of a mess than it already is.

There is an example of usage of reraise in parsing/location.ml: you bind it locally as an external, and then you use it. Not the most pleasant thing, but that should only be necessary for low-level stuff. (For high-level stuff I would just call `Fun.protect` instead of having to think about all that.) But here you are trying to write code that is robust against asynchronous exceptions, so you have to think about low-level details, and `reraise` appears to be the best option -- if I understand correctly, you can't afford a function call before `unlock` to store the backtrace.

It may also be correct to use `get_raw_backtrace` *after* the call to `unlock` and to use `raise_with_backtrace` immediatley after that; that way would only use functions exposed in the stdlib, but it is also slower (you save the current backtrace and restore it, instead of just leaving it in place).

Reply:
I thought `raise e` was automatically understood by the compiler as meaning `reraise e` in this case.

Reply:
well, it's done